File: android/app/src/main/java/dev/droplet/app/mesh/TrustList.kt
package dev.droplet.app.mesh

import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.net.Inet4Address
import java.net.Inet6Address
import java.net.InetAddress

/**
 * The peers this phone talks to (docs/mesh.md §3), as the Linux agent keeps
 * them: keyed by fingerprint, each with its full certificate.
 *
 * - "roster": vouched for by the hub, replaced wholesale by each roster
 *   fetch, kept while the hub is down.
 * - "paired": paired directly by the owner; survives roster changes and
 *   goes only when unpaired.
 */
class TrustList(private val file: File, private val ownFp: String) {
    data class Entry(
        val id: String,
        val name: String,
        val fp: String,
        val certPem: String,
        val source: String,
        val lan: List<String> = emptyList(),
        val port: Int? = null,
        val tailnetIp: String? = null,
        val os: String = "",
        val caps: List<String> = emptyList(),
        /** The hub whose roster lists it: messages may go through that hub. */
        val hub: String = "",
        val added: Long = System.currentTimeMillis() / 1000,
    ) {
        fun toJson(): JSONObject = JSONObject().put("id", id).put("name", name).put("fp", fp).put("cert_pem", certPem)
            .put("source", source).put("lan", JSONArray(lan)).put("port", port ?: JSONObject.NULL)
            .put("tailnet_ip", tailnetIp ?: JSONObject.NULL).put("os", os).put("caps", JSONArray(caps))
            .put("hub", hub).put("added", added)
    }

    /** Called (outside the lock) when the set of trusted certificates changed. */
    @Volatile var onChange: (() -> Unit)? = null

    private val lock = Any()
    private var peers = LinkedHashMap<String, Entry>()

    init {
        load()
    }

    private fun load() {
        val raw = runCatching { JSONObject(file.readText()) }.getOrNull()?.optJSONObject("peers") ?: return
        for (fp in raw.keys()) {
            val e = raw.optJSONObject(fp) ?: continue
            try {
                val entry = makeEntry(
                    peerId = e.optString("id"), name = e.optString("name"), certPem = e.optString("cert_pem"),
                    source = e.optString("source"), lan = strings(e.optJSONArray("lan")), port = e.opt("port"),
                    tailnetIp = e.optString("tailnet_ip").takeIf { !e.isNull("tailnet_ip") }, os = e.optString("os"),
                    caps = strings(e.optJSONArray("caps")), fp = fp, hub = e.optString("hub"),
                ).copy(added = e.optLong("added").takeIf { it > 0 } ?: (System.currentTimeMillis() / 1000))
                if (entry.fp != ownFp) peers[entry.fp] = entry
            } catch (x: IllegalArgumentException) {
                // a damaged entry is dropped, never trusted
            }
        }
    }

    private fun save() {
        val out = JSONObject()
        peers.values.forEach { out.put(it.fp, it.toJson()) }
        file.parentFile?.mkdirs()
        MeshIdentity.writePrivate(file, JSONObject().put("v", 1).put("peers", out).toString(1).toByteArray())
    }

    private fun changed(certs: Boolean) {
        save()
        if (certs) onChange?.let { cb -> Thread { cb() }.start() }
    }

    // --- reading -------------------------------------------------------------

    fun get(fp: String?): Entry? = synchronized(lock) { peers[fp ?: return null] }

    fun isTrusted(fp: String): Boolean = get(fp) != null

    fun all(): List<Entry> = synchronized(lock) { peers.values.sortedWith(compareBy({ it.name.lowercase() }, { it.fp })) }

    /** Peers matching a name (any case), an id, or a fingerprint prefix of 8+ hex. */
    fun find(query: String): List<Entry> {
        val q = query.trim()
        val ql = q.lowercase().replace(":", "")
        val list = all()
        val exact = list.filter { it.id == ql || it.fp == ql || it.name.equals(q, ignoreCase = true) }
        if (exact.isNotEmpty()) return exact
        return if (ql.length >= 8) list.filter { it.fp.startsWith(ql) } else emptyList()
    }

    // --- changing -------------------------------------------------------------

    /** Trusts a directly paired peer. A re-pair of the same device id replaces its old certificate. */
    fun addPaired(entry: Entry) {
        require(entry.source == SOURCE_PAIRED)
        require(entry.fp != ownFp) { "that's this device" }
        synchronized(lock) {
            peers.entries.removeAll { it.value.id == entry.id && it.value.source == SOURCE_PAIRED && it.key != entry.fp }
            val old = peers[entry.fp]
            peers[entry.fp] = if (old == null) entry else entry.copy(
                lan = cleanAddresses(entry.lan + old.lan), tailnetIp = entry.tailnetIp ?: old.tailnetIp,
                port = entry.port ?: old.port)
            changed(true)
        }
    }

    fun remove(fp: String): Entry? = synchronized(lock) {
        peers.remove(fp)?.also { changed(true) }
    }

    /** Makes the roster peers exactly [entries] (from hub [hubId]); paired peers stay and learn addresses. Returns (added, removed). */
    fun syncRoster(entries: List<Entry>, hubId: String): Pair<Int, Int> = synchronized(lock) {
        val fresh = entries.filter { it.fp != ownFp }.associateBy { it.fp }
        val before = peers.mapValues { it.value.certPem }
        var removed = 0
        for ((fp, e) in peers.entries.toList()) {
            if (e.source == SOURCE_ROSTER && fp !in fresh) {
                peers.remove(fp)
                removed++
            } else if (e.source == SOURCE_PAIRED && fp !in fresh && e.hub == hubId) {
                peers[fp] = e.copy(hub = "")   // still paired, but that hub no longer knows it
            }
        }
        var added = 0
        for ((fp, e) in fresh) {
            val old = peers[fp]
            peers[fp] = when {
                old == null -> e.also { added++ }
                // direct pairing is the owner's decision: it stays "paired" and learns where the peer is
                old.source == SOURCE_PAIRED -> old.copy(lan = cleanAddresses(e.lan + old.lan), port = e.port ?: old.port,
                    tailnetIp = e.tailnetIp ?: old.tailnetIp, os = e.os.ifEmpty { old.os },
                    caps = e.caps.ifEmpty { old.caps }, hub = e.hub)
                else -> e.copy(added = old.added, lan = cleanAddresses(e.lan + old.lan))
            }
        }
        changed(peers.mapValues { it.value.certPem } != before)
        added to removed
    }

    /** Records what an authenticated peer said about itself, or where it answered. */
    fun learn(fp: String, address: String? = null, port: Int? = null, name: String? = null, peerId: String? = null,
              os: String? = null, caps: List<String>? = null, tailnet: Boolean = false) = synchronized(lock) {
        val e = peers[fp] ?: return@synchronized
        var n = e
        if (address != null) {
            n = if (tailnet) n.copy(tailnetIp = cleanAddresses(listOf(address), 1).firstOrNull() ?: n.tailnetIp)
            else n.copy(lan = cleanAddresses(listOf(address) + n.lan))
        }
        if (port != null && port in 1..65535) n = n.copy(port = port)
        if (n.source == SOURCE_PAIRED) {
            // the hub names roster peers; a paired peer names itself
            if (!name.isNullOrBlank()) n = n.copy(name = cleanName(name, n.name))
            if (peerId != null && PEER_ID.matches(peerId)) n = n.copy(id = peerId)
        }
        if (!os.isNullOrBlank()) n = n.copy(os = cleanName(os).take(20))
        if (caps != null) n = n.copy(caps = cleanCaps(caps))
        if (n != e) {
            peers[fp] = n
            save()
        }
    }

    companion object {
        const val SOURCE_ROSTER = "roster"
        const val SOURCE_PAIRED = "paired"
        val PEER_ID = Regex("^[0-9a-f]{8,64}$")
        val FINGERPRINT = Regex("^[0-9a-f]{64}$")
        private const val MAX_LAN = 6

        private fun strings(a: JSONArray?): List<String> = (0 until (a?.length() ?: 0)).mapNotNull { a!!.opt(it) as? String }

        /** A checked entry. Throws IllegalArgumentException when the id, certificate or fingerprint is wrong. */
        fun makeEntry(peerId: String?, name: String?, certPem: String?, source: String, lan: List<String> = emptyList(),
                      port: Any? = null, tailnetIp: String? = null, os: String? = "", caps: List<String> = emptyList(),
                      fp: String? = null, hub: String? = null): Entry {
            require(source == SOURCE_ROSTER || source == SOURCE_PAIRED) { "bad source" }
            require(peerId != null && PEER_ID.matches(peerId)) { "bad peer id" }
            val der = MeshIdentity.pemToDer(certPem)
            val real = MeshIdentity.sha256Hex(der)
            if (fp != null) require(fp.replace(":", "").trim().lowercase() == real) { "the fingerprint doesn't match the certificate" }
            val p = (port as? Number)?.toInt()?.takeIf { it in 1..65535 && port !is Double }
            return Entry(
                id = peerId, name = cleanName(name, peerId), fp = real, certPem = MeshIdentity.toPem(der), source = source,
                lan = cleanAddresses(lan), port = p, tailnetIp = tailnetIp?.let { cleanAddresses(listOf(it), 1).firstOrNull() },
                os = cleanName(os).take(20), caps = cleanCaps(caps),
                hub = hub?.takeIf { PEER_ID.matches(it) } ?: "",
            )
        }

        fun cleanName(v: String?, fallback: String = ""): String {
            val name = (v ?: "").split(Regex("\\s+")).filter { it.isNotEmpty() }.joinToString(" ")
                .filter { !Character.isISOControl(it) && Character.getType(it) != Character.FORMAT.toInt() }.take(64)
            return name.ifEmpty { fallback }
        }

        fun cleanCaps(v: List<String>): List<String> =
            v.map { it.trim() }.filter { c -> c.isNotEmpty() && c.length <= 20 && c.all { it.isLetterOrDigit() && it.code < 128 || it == '-' } }
                .toSortedSet().toList()

        /** An IP literal, never a DNS lookup. */
        fun ipLiteral(s: String?): InetAddress? {
            val t = s?.trim()?.substringBefore('%')?.removePrefix("[")?.removeSuffix("]") ?: return null
            val v4 = Regex("^(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})$").matchEntire(t)
            if (v4 != null) {
                if (v4.groupValues.drop(1).any { it.toInt() > 255 }) return null
                return runCatching { InetAddress.getByName(t) }.getOrNull()
            }
            if (':' in t && t.all { it.isDigit() || it in 'a'..'f' || it in 'A'..'F' || it == ':' || it == '.' }) {
                return runCatching { InetAddress.getByName(t) }.getOrNull() as? Inet6Address
            }
            return null
        }

        /** The usual text form: dotted IPv4, compressed IPv6 ("fd00::1"), as Python writes them. */
        fun text(ip: InetAddress): String {
            if (ip is Inet4Address) return ip.hostAddress!!
            val b = ip.address
            val groups = (0 until 8).map { ((b[it * 2].toInt() and 0xff) shl 8) or (b[it * 2 + 1].toInt() and 0xff) }
            // the longest run of two or more zero groups becomes "::" (the first, if tied)
            var best = -1
            var bestLen = 1
            var i = 0
            while (i < 8) {
                if (groups[i] == 0) {
                    var j = i
                    while (j < 8 && groups[j] == 0) j++
                    if (j - i > bestLen) { best = i; bestLen = j - i }
                    i = j
                } else i++
            }
            val hex = groups.map { Integer.toHexString(it) }
            return if (best < 0) hex.joinToString(":")
            else hex.subList(0, best).joinToString(":") + "::" + hex.subList(best + bestLen, 8).joinToString(":")
        }

        fun cleanAddresses(items: List<String>, limit: Int = MAX_LAN): List<String> {
            val out = ArrayList<String>()
            for (a in items) {
                val ip = ipLiteral(a) ?: continue
                if (ip.isAnyLocalAddress || ip.isMulticastAddress || (ip is Inet6Address && ip.isLinkLocalAddress)) continue
                val s = text(ip)
                if (s !in out) out.add(s)
            }
            return out.take(limit)
        }
    }
}

File: windows-net/src/Droplet.Core/Mesh/TrustList.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using Droplet.Core.Common;

namespace Droplet.Core.Mesh;

/// <summary>Why a peer is trusted.</summary>
public static class TrustSource
{
    /// <summary>Vouched for by the hub, through its roster. Replaced by each roster fetch.</summary>
    public const string Roster = "roster";

    /// <summary>Paired directly by the owner. Only goes when unpaired.</summary>
    public const string Paired = "paired";
}

/// <summary>One trusted peer (docs/mesh.md §3). The certificate is kept in full.</summary>
public sealed record TrustEntry
{
    /// <summary>Its peer id.</summary>
    [JsonPropertyName("id")] public required string Id { get; init; }

    /// <summary>Its name.</summary>
    [JsonPropertyName("name")] public required string Name { get; init; }

    /// <summary>Its certificate's fingerprint.</summary>
    [JsonPropertyName("fp")] public required string Fp { get; init; }

    /// <summary>Its certificate, PEM.</summary>
    [JsonPropertyName("cert_pem")] public required string CertPem { get; init; }

    /// <summary><see cref="TrustSource"/>.</summary>
    [JsonPropertyName("source")] public required string Source { get; init; }

    /// <summary>LAN addresses it was seen at, newest first.</summary>
    [JsonPropertyName("lan")] public List<string> Lan { get; init; } = [];

    /// <summary>Its mesh port.</summary>
    [JsonPropertyName("port")] public int? Port { get; init; }

    /// <summary>Its tailnet address.</summary>
    [JsonPropertyName("tailnet_ip")] public string? TailnetIp { get; init; }

    /// <summary>android, windows or linux.</summary>
    [JsonPropertyName("os")] public string Os { get; init; } = "";

    /// <summary>What it can do.</summary>
    [JsonPropertyName("caps")] public List<string> Caps { get; init; } = [];

    /// <summary>The hub whose roster lists it: messages may go through that hub.</summary>
    [JsonPropertyName("hub")] public string Hub { get; init; } = "";

    /// <summary>When it was added, Unix seconds.</summary>
    [JsonPropertyName("added")] public long Added { get; init; }

    /// <summary>
    /// A checked entry. Throws <see cref="FormatException"/> when the id, certificate or
    /// fingerprint is wrong (a damaged or forged entry is never trusted).
    /// </summary>
    public static TrustEntry Make(string? peerId, string? name, string? certPem, string source, IEnumerable<string?>? lan = null,
        long? port = null, string? tailnetIp = null, string? os = null, IEnumerable<string>? caps = null, string? fp = null,
        string? hub = null)
    {
        if (source is not (TrustSource.Roster or TrustSource.Paired))
        {
            throw new FormatException($"bad source {source}");
        }
        if (peerId is null || !MeshIdentity.PeerIdPattern().IsMatch(peerId))
        {
            throw new FormatException("bad peer id");
        }
        var der = Certificates.PemToDer(certPem);
        var real = Fingerprint.Of(der);
        if (fp is not null && Fingerprint.Normalize(fp) != real)
        {
            throw new FormatException("the fingerprint doesn't match the certificate");
        }
        var tip = Addresses.Clean(tailnetIp is null ? [] : [tailnetIp], 1);
        return new TrustEntry
        {
            Id = peerId,
            Name = Clean.Name(name, peerId),
            Fp = real,
            CertPem = Certificates.ToPem(der),
            Source = source,
            Lan = Addresses.Clean(lan),
            Port = Clean.Port(port),
            TailnetIp = tip.Count > 0 ? tip[0] : null,
            Os = Truncate(Clean.Name(os), 20),
            Caps = Clean.Caps(caps),
            Hub = hub is not null && MeshIdentity.PeerIdPattern().IsMatch(hub) ? hub : "",
            Added = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
        };
    }

    internal static string Truncate(string s, int n) => s.Length > n ? s[..n] : s;
}

/// <summary>
/// The trust list: the peers this device talks to (docs/mesh.md §3). Kept as JSON
/// (<c>{"v":1,"peers":{"&lt;fp&gt;": entry}}</c>, the reference's layout), keyed by
/// fingerprint. Thread-safe.
/// </summary>
public sealed class TrustList
{
    static readonly JsonSerializerOptions Options = new(Json.Indented)
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.Never,
    };

    readonly string path;
    readonly string ownFp;
    readonly Lock gate = new();
    Dictionary<string, TrustEntry> peers = [];

    /// <summary>Loads the list at <paramref name="path"/>; this device's own fingerprint is never trusted.</summary>
    public TrustList(string path, string ownFp)
    {
        this.path = path;
        this.ownFp = ownFp;
        Load();
    }

    /// <summary>
    /// Raised (outside any lock) after the set of trusted certificates changes: links to a
    /// peer no longer trusted must close.
    /// </summary>
    public event Action? CertificatesChanged;

    sealed record FileForm([property: JsonPropertyName("v")] int V, [property: JsonPropertyName("peers")] Dictionary<string, JsonElement>? Peers);

    void Load()
    {
        var text = AtomicFile.TryReadText(path);
        if (text is null)
        {
            return;
        }
        FileForm? raw;
        try
        {
            raw = JsonSerializer.Deserialize<FileForm>(text);
        }
        catch (JsonException)
        {
            return;
        }
        var loaded = new Dictionary<string, TrustEntry>();
        foreach (var (fp, e) in raw?.Peers ?? [])
        {
            try
            {
                var stored = e.Deserialize<TrustEntry>();
                if (stored is null)
                {
                    continue;
                }
                // re-checked, so a damaged entry is dropped, never trusted
                var entry = TrustEntry.Make(stored.Id, stored.Name, stored.CertPem, stored.Source, stored.Lan, stored.Port,
                    stored.TailnetIp, stored.Os, stored.Caps, fp, stored.Hub) with
                { Added = stored.Added > 0 ? stored.Added : DateTimeOffset.UtcNow.ToUnixTimeSeconds() };
                if (entry.Fp != ownFp)
                {
                    loaded[entry.Fp] = entry;
                }
            }
            catch (Exception ex) when (ex is JsonException or FormatException or InvalidOperationException)
            {
            }
        }
        peers = loaded;
    }

    void Save()
    {
        var data = JsonSerializer.SerializeToUtf8Bytes(new { v = 1, peers }, Options);
        AtomicFile.Write(path, data);
    }

    // --- reading ---------------------------------------------------------------------

    /// <summary>The entry for a fingerprint, or null.</summary>
    public TrustEntry? Get(string? fp)
    {
        if (fp is null)
        {
            return null;
        }
        lock (gate)
        {
            return peers.GetValueOrDefault(fp);
        }
    }

    /// <summary>Every trusted peer, by name.</summary>
    public List<TrustEntry> All()
    {
        lock (gate)
        {
            return [.. peers.Values.OrderBy(e => e.Name.ToLowerInvariant(), StringComparer.Ordinal).ThenBy(e => e.Fp, StringComparer.Ordinal)];
        }
    }

    /// <summary>Peers matching a name (case-insensitive), an id, a fingerprint, or a fingerprint prefix of 8+ hex.</summary>
    public List<TrustEntry> Find(string? query)
    {
        var q = (query ?? "").Trim();
        var ql = q.ToLowerInvariant();
        List<TrustEntry> all;
        lock (gate)
        {
            all = [.. peers.Values];
        }
        var exact = all.Where(e => e.Id == ql || e.Fp == ql || string.Equals(e.Name, q, StringComparison.OrdinalIgnoreCase)).ToList();
        if (exact.Count > 0)
        {
            return exact;
        }
        if (ql.Length >= 8)
        {
            var prefix = ql.Replace(":", "", StringComparison.Ordinal);
            return all.Where(e => e.Fp.StartsWith(prefix, StringComparison.Ordinal)).ToList();
        }
        return [];
    }

    // --- changing ----------------------------------------------------------------------

    /// <summary>Trusts a directly paired peer. A re-pair of the same peer id replaces its old certificate.</summary>
    public void AddPaired(TrustEntry entry)
    {
        ArgumentNullException.ThrowIfNull(entry);
        if (entry.Source != TrustSource.Paired)
        {
            throw new ArgumentException("not a paired entry", nameof(entry));
        }
        lock (gate)
        {
            if (entry.Fp == ownFp)
            {
                throw new InvalidOperationException("that's this device");
            }
            foreach (var (fp, e) in peers.ToList())
            {
                if (e.Id == entry.Id && e.Source == TrustSource.Paired && fp != entry.Fp)
                {
                    peers.Remove(fp);
                }
            }
            if (peers.TryGetValue(entry.Fp, out var old))
            {
                entry = entry with
                {
                    Lan = Addresses.Clean(entry.Lan.Concat(old.Lan)),
                    TailnetIp = entry.TailnetIp ?? old.TailnetIp,
                };
            }
            peers[entry.Fp] = entry;
            Save();
        }
        CertificatesChanged?.Invoke();
    }

    /// <summary>Stops trusting a peer. Returns what was removed.</summary>
    public TrustEntry? Remove(string fp)
    {
        TrustEntry? e;
        lock (gate)
        {
            if (!peers.Remove(fp, out e))
            {
                return null;
            }
            Save();
        }
        CertificatesChanged?.Invoke();
        return e;
    }

    /// <summary>
    /// Makes the roster peers exactly <paramref name="entries"/> (from hub
    /// <paramref name="hubId"/>). Paired peers stay, and learn fresher addresses from it.
    /// Returns (added, removed).
    /// </summary>
    public (int Added, int Removed) SyncRoster(IReadOnlyList<TrustEntry> entries, string hubId)
    {
        ArgumentNullException.ThrowIfNull(entries);
        int added = 0, removed = 0;
        bool changed;
        lock (gate)
        {
            var incoming = new Dictionary<string, TrustEntry>();
            foreach (var e in entries)
            {
                if (e.Fp != ownFp)
                {
                    incoming[e.Fp] = e;
                }
            }
            var before = peers.ToDictionary(p => p.Key, p => p.Value.CertPem);
            foreach (var (fp, e) in peers.ToList())
            {
                if (e.Source == TrustSource.Roster && !incoming.ContainsKey(fp))
                {
                    peers.Remove(fp);
                    removed++;
                }
                else if (e.Source == TrustSource.Paired && !incoming.ContainsKey(fp) && e.Hub == hubId)
                {
                    peers[fp] = e with { Hub = "" }; // still paired, but that hub no longer knows it
                }
            }
            foreach (var (fp, e) in incoming)
            {
                if (!peers.TryGetValue(fp, out var old))
                {
                    peers[fp] = e;
                    added++;
                }
                else if (old.Source == TrustSource.Paired)
                {
                    // direct pairing is the owner's own decision: it stays "paired", and
                    // learns where the peer is from the hub
                    peers[fp] = old with
                    {
                        Lan = Addresses.Clean(e.Lan.Concat(old.Lan)),
                        Port = e.Port ?? old.Port,
                        TailnetIp = e.TailnetIp ?? old.TailnetIp,
                        Os = e.Os.Length > 0 ? e.Os : old.Os,
                        Caps = e.Caps.Count > 0 ? e.Caps : old.Caps,
                        Hub = e.Hub,
                    };
                }
                else
                {
                    peers[fp] = e with { Added = old.Added, Lan = Addresses.Clean(e.Lan.Concat(old.Lan)) };
                }
            }
            var after = peers.ToDictionary(p => p.Key, p => p.Value.CertPem);
            changed = before.Count != after.Count || before.Any(b => !after.TryGetValue(b.Key, out var pem) || pem != b.Value);
            Save();
        }
        if (changed)
        {
            CertificatesChanged?.Invoke();
        }
        return (added, removed);
    }

    /// <summary>Records what an authenticated peer said about itself, or where it answered.</summary>
    public void Learn(string fp, string? address = null, long? port = null, string? name = null, string? peerId = null,
        string? os = null, IEnumerable<string>? caps = null, bool tailnet = false)
    {
        lock (gate)
        {
            if (!peers.TryGetValue(fp, out var e))
            {
                return;
            }
            var before = JsonSerializer.Serialize(e);
            if (!string.IsNullOrEmpty(address))
            {
                if (tailnet)
                {
                    var t = Addresses.Clean([address], 1);
                    e = e with { TailnetIp = t.Count > 0 ? t[0] : e.TailnetIp };
                }
                else
                {
                    e = e with { Lan = Addresses.Clean(new[] { address }.Concat(e.Lan)) };
                }
            }
            if (Clean.Port(port) is { } p)
            {
                e = e with { Port = p };
            }
            if (e.Source == TrustSource.Paired)
            {
                // the hub names roster peers; a paired peer names itself
                if (!string.IsNullOrEmpty(name))
                {
                    e = e with { Name = Clean.Name(name, e.Name) };
                }
                if (peerId is not null && MeshIdentity.PeerIdPattern().IsMatch(peerId))
                {
                    e = e with { Id = peerId };
                }
            }
            if (!string.IsNullOrEmpty(os))
            {
                e = e with { Os = TrustEntry.Truncate(Clean.Name(os), 20) };
            }
            if (caps is not null)
            {
                e = e with { Caps = Clean.Caps(caps) };
            }
            if (JsonSerializer.Serialize(e) != before)
            {
                peers[fp] = e;
                Save();
            }
        }
    }
}
